lib/apiservers/engine/backends/container.go

```diff
@@ -506,3 +663,185 @@ func (c *Container) imageExist(imageID string) (storeName string, err error) {
+// clStdin, clStdout, clStderr are the hijacked connection
+// keys are the keys that are used for detaching streams without closing them
+func attachStreams(name string, tty, stdinOnce bool, clStdin io.ReadCloser, clStdout, clStderr io.Writer, keys []byte) error {
+	var wg sync.WaitGroup
+
+	errors := make(chan error, 3)
+	//FIXME: Swagger will timeout on us.  We need to either have an infinite timeout or the timeout should
+	// start after some inactivity?
+	inContext, inCancel := context.WithTimeout(context.Background(), 2*time.Hour)
+	outContext, outCancel := context.WithTimeout(context.Background(), 2*time.Hour)
+	errContext, errCancel := context.WithTimeout(context.Background(), 2*time.Hour)
+
+	if clStdin != nil {
+		wg.Add(1)
+		// Pipe for stdin so we can interject and watch the input streams for detach keys.
+		stdinReader, stdinWriter := io.Pipe()
+
+		defer clStdin.Close()
+		defer stdinReader.Close()
+
+		transport := httptransport.New(PortLayerServer(), "/", []string{"http"})
+		plClient := client.New(transport, nil)
+		transport.Consumers["application/octet-stream"] = httpkit.ByteStreamConsumer()
+		transport.Producers["application/octet-stream"] = httpkit.ByteStreamProducer()
+
+		// Swagger wants an io.reader so give it the reader pipe.  Also, the swagger call
+		// to set the stdin is synchronous so we need to run in a goroutine
+		go func() {
+			setStdinParams := interaction.NewContainerSetStdinParamsWithContext(inContext).WithID(name)
+			setStdinParams = setStdinParams.WithRawStream(stdinReader)
+			plClient.Interaction.ContainerSetStdin(setStdinParams)
+		}()
+
+		// Copy the stdin from the CLI and write to a pipe.  We need to do this so we can
+		// watch the stdin stream for the detach keys.
+		go func() {
+			var err error
+			if tty {
+				_, err = attach.CopyEscapable(stdinWriter, clStdin, keys)
+			} else {
+				_, err = io.Copy(stdinWriter, clStdin)
+			}
+
+			if err != nil {
+				log.Errorf(err.Error())
+			}
+			errors <- err
+
+			if stdinOnce && !tty {
+				// Close the stdin connection.  Mimicing Docker's behavior.
+				// FIXME: If we close this stdin connection and request it again,
+				// does the port layer reconnects the stdin?  Is this diff from
+				// current Docker behavior?
+				stdinReader.Close()
+			} else {
+				// Shutdown the client's request for stdout, stderr
+				errCancel()
+				outCancel()
+			}
+			wg.Done()
+		}()
+	}
+
+	if clStdout != nil {
+		wg.Add(1)
+		transport := httptransport.New(PortLayerServer(), "/", []string{"http"})
+		plClient := client.New(transport, nil)
+		transport.Consumers["application/octet-stream"] = httpkit.ByteStreamConsumer()
+		transport.Producers["application/octet-stream"] = httpkit.ByteStreamProducer()
+
+		// Swagger -> pipewriter.  Synchronous, blocking call
+		go func() {
+			getStdoutParams := interaction.NewContainerGetStdoutParamsWithContext(outContext).WithID(name)
```

Comment:
Same go routine comment.  Please take this out of the routine and do this in the parent function.